### PR TITLE
refactoring

### PR DESCRIPTION
**Pull request by bdehri**
* add tests to iam package
* refactor reconcilers

### CODE REVIEW

File: pkg/key/key.go [hunk 1]
```diff
@@ -147,3 +149,21 @@ func GetBaseDomain(ctx context.Context, ctrlClient client.Client, clusterName, n
 
 	return baseDomain, nil
 }
+
+func GetAWSAccountID(awsClusterRoleIdentity *capa.AWSClusterRoleIdentity) (string, error) {
```
Comment (AndiDog) [inline pkg/key/key.go:153]: Let's use opsctl's battle-tested function `extractAwsAccountIDfromARN` which uses the SDK for parsing

**Review by AndiDog — COMMENTED** [review]
Not a full review yet

File: controllers/awsmachinepool_controller.go [hunk 2]
```diff
@@ -117,96 +115,64 @@ func (r *AWSMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
-			err = patchHelper.Patch(ctx, awsMachinePool)
-			if err != nil {
-				logger.Error(err, "failed to add finalizer on AWSMachinePool")
-				return ctrl.Result{}, errors.WithStack(err)
-			}
-			logger.Info("successfully added finalizer to AWSMachinePool", "finalizer_name", iam.NodesRole)
-		}
+	return ctrl.Result{
+		Requeue:      true,
+		RequeueAfter: time.Minute * 5,
```
Comment (fiunchinho) [inline controllers/awsmachinepool_controller.go:153]: Why are we requeuing?

File: controllers/awsmachinepool_controller.go [hunk 3]
```diff
@@ -117,96 +115,64 @@ func (r *AWSMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
 		}
+		logger.Info("successfully added finalizer to AWSMachinePool", "finalizer_name", iam.NodesRole)
+	}
+
+	err := iamService.ReconcileRole()
+	if err != nil {
+		return ctrl.Result{}, errors.WithStack(err)
 	}
 
 	return ctrl.Result{
```
Comment (fiunchinho) [inline controllers/awsmachinepool_controller.go:178]: Same as above, do we need to requeue when the reconciliation was successfull?

File: controllers/awsmachinetemplate_controller.go [hunk 4]
```diff
@@ -136,201 +134,217 @@ func (r *AWSMachineTemplateReconciler) Reconcile(ctx context.Context, req ctrl.R
...
 
-			if controllerutil.ContainsFinalizer(awsCluster, key.FinalizerName(iam.ControlPlaneRole)) {
-				patchHelper, err := patch.NewHelper(awsCluster, r.Client)
+	if !roleUsed {
+		err = iamService.DeleteRole()
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if role == iam.ControlPlaneRole {
+			if r.EnableKiamRole {
```
Comment (fiunchinho) [inline controllers/awsmachinetemplate_controller.go:155]: Didn't we remove `kiam` in CAPA? If that's the case, why not remove all logic related to `kiam` from the operator?

Comment (bdehri) [reply controllers/awsmachinetemplate_controller.go:155]: I tried it but they are somehow still required (I was not able to create a cluster with them) but we might rename respective parts

File: controllers/awsmachinepool_controller.go (hunk 2 above)
Comment (bdehri) [reply controllers/awsmachinepool_controller.go:153]: It was like this before so I assumed its there for a purpose.

File: controllers/awsmachinepool_controller.go [hunk 5]
```diff
@@ -117,96 +115,64 @@ func (r *AWSMachinePoolReconciler) Reconcile(ctx context.Context, req ctrl.Reque
...
-				return ctrl.Result{}, errors.WithStack(err)
-			}
+	// remove finalizer from AWSMachinePool
+	if controllerutil.ContainsFinalizer(awsMachinePool, key.FinalizerName(iam.NodesRole)) {
+		patchHelper, err := patch.NewHelper(awsMachinePool, r.Client)
+		if err != nil {
+			return ctrl.Result{}, errors.WithStack(err)
 		}
-
-		// remove finalizer from AWSCluster
```
[the remark this answers is not in the document]
Comment (bdehri) [reply controllers/awsmachinepool_controller.go:132]: I dont understand how having finalizer ensures that behavior as roles are deleted first and we dont check if capa already reconciled deletion of AWSCluster.

[the remark this answers is not in the document]
Comment (AndiDog) [reply controllers/awsmachinepool_controller.go:132]: We can do it like in aws-vpc-operator and wait until CAPA has deleted its stuff and removed the finalizer:

```
//
// Wait for CAPA to delete load balancer before we delete VPC, subnets and route tables.
//
if controllerutil.ContainsFinalizer(awsCluster, capa.ClusterFinalizer) {
```

**Comment by bdehri** [comment]
I reverted the latest change as it introduced deadlock between awsMachinePool and awsCluster.

**Comment by AndiDog** [comment]
> I reverted the latest change as it introduced deadlock between awsMachinePool and awsCluster.

Unfortunately, I don't have a great idea to solve this. We need the machine pool name for deletion, so we can't remove the `AWSMachinePool` finalizer first either. Let's go ahead as-is and maybe we'll find a solution later. This is not a large problem, but leads to false error logs during deletion (`AWSCluster` reconciler failing to reconcile S3 bucket IAM policy once the IAM roles are missing).

I'll review this one more time.

**Review by AndiDog — APPROVED** [review]

**Comment by mnitchev** [comment]
Towards giantswarm/roadmap#2113

**Pull request merged**